### ISSUE TITLE
Unyson problem

### ISSUE THREAD

**Issue by ziuzea87** [ThemeFuse/Unyson#138]
I get this message when updating extensions. Why?

Downloading the Sidebars extension...

Github error: API rate limit exceeded for 50.87.248.69. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)

Downloading the Update extension...

Github error: API rate limit exceeded for 50.87.248.69. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)

**Comment by 2Fwebd** [comment]
Hello, 
This problem, is more than than problematic...This solution is not reasonable for premium themes. I mean if customers pay more than 50$ for a theme and have to wait an hour for every extension, it's pretty awful, no ?
Then, I guess the best way is to include directly the extensions supported by the theme directly in the framework-customization/extensions folder, It works for all of them except for the page-builder. Because there is a relation (parent) with the shortcodes one no ? So I get the error : cannot define again the extension page-builder (already defined in shortcodes... and the folder is there but empty).
Well, does it make sense ? I'm just looking for the best way to load the extensions .. :) Sorry I'm not that good developer but I try to learn. By the way, as I already said before you've made an awesome work here, I advance too fast of my theme development process !!

**Comment by moldcraft** [comment]
If you add extensions pre-installed in `framework-customization/extensions/` your theme will not pass theme-check and marketplace approval.

----

We can add a config parameter for the `github-update` extension to disable the cache reset when the user click the "Check Now" button. I think this may help a lot to prevent "API rate limit exceeded" error.

What do you think?

**Comment by 2Fwebd** [comment]
I forgot that detail...Well, I spend sometime on the Google trying to understand what happens and : 
> For requests using Basic Authentication or OAuth, you can **make up to 5,000 requests per hour**. For unauthenticated requests, the rate limit allows you to make up to 60 requests per hour. 

That would be awesome ! Problem is the OAuth no ? it will becomes public ? 

Then, I found this : (http://stackoverflow.com/questions/16732103/increase-github-api-limit)
> You have to contact GitHub support to have your ratelimit increased, contrary to what spuder claims. They'll raise it for you.
As for not sharing your client_id or client_secret, spuder has the right answer there. 

And (http://stackoverflow.com/questions/13394077/is-there-a-way-to-increase-the-api-rate-limit-or-to-bypass-it-altogether-for-git) 

What do you think ? let me know if I can help.  Thank you very much again!

**Comment by 2Fwebd** [comment]
Any news or I am the only one to get this issue every time I try to download an extension ? :P Let me know if I can help you.

**Comment by moldcraft** [comment]
> Any news?

not yet

> or I am the only one to get this issue every time I try to download an extension ?

Do you mean literally **every time**, or you press often the [Check Again](http://static.md/4467f83abc7db67bc4e85cb16316c94a.png) button?

**Comment by 2Fwebd** [comment]
Well no, I just did a test and sorry you are right when I get the error **if I wait one hour I can install a few ones but (5 actually)** then the error comes back. I never pressed the Check Again button. 

In fact, there is no way to make an authentification for Unyson ? **That would be perfect, 6000 per hour should be enough, no ?**

**Comment by moldcraft** [comment]
> https://developer.github.com/v3/#rate-limiting
> For requests using Basic Authentication or OAuth, you can make up to 5,000 requests per hour. For unauthenticated requests, the rate limit allows you to make up to 60 requests per hour. Unauthenticated requests are associated with your IP address, and not the user making requests.

If we will hardcode our token in the framework core and make API requests with it, it will limit all Unyson users to 5000 pe hour. So this is not a good solution.

**Comment by 2Fwebd** [comment]
Sorry I thought it was 5000 requests per user....In that case yes it is not the good solution. 

Yet, I don't understand how I pay away these requests. There is maybe a problem there. 

I just did another test :
* Went to one of my dev site,
* Unyson page in Wordpress
* Clicked on "Download" on the feedback extension that I don't use.
* Got still the same message : 

> "Github error: API rate limit exceeded for 213.251.182.10. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)"

I wasn't on this website since more than 12 hours...So I didn't do any request...

So, the best way will be to delete those automatic requests ?

**Comment by moldcraft** [comment]
We thought it will not be a big problem but I seems critical.

I think we will do an intermediary site that makes requests to github using our OAuth token and cache them, and all Unyson users will make unlimited requests to that site.

```text
[Unyson User] ---> [Cached API Requests] ---> [Github API]
```

**Comment by 2Fwebd** [comment]
That should be very good ! Thank you @moldcraft  !!

**Comment by 2Fwebd** [comment]
Well, I get this sometime : 
![starry___just_another_wordpress_site](https://cloud.githubusercontent.com/assets/5446124/5644545/ee17e984-9664-11e4-8487-f3a83ff2d146.png)
On the frontend ... Have you already seen this before ?

**Comment by moldcraft** [comment]
> On the frontend ... Have you already seen this before ?

The framework uses `FW_Flash_Messages::add()` to display notice messages, they are displayed both in backend and frontend.

But don't worry, the update extension is activated only if current user is administrator and has permission to update themes/plugins, so simple users will not see this message. These messages are stored in `$_SESSION` so only you will see them.

**Comment by danyj** [comment]
do you guys need dedicated server for this? It is obvious that many devs are being pulled here since we all need reliable framework but this API thing can kill us all. If you need dedicated server let me know.

**Comment by moldcraft** [comment]
No, thank you. Everything is ready. I'am releasing a new version now *(2.1.13)*

**Comment by moldcraft** [comment]
Please update to v2.1.13 and let us know if the `Github API rate limit exceeded` is still there

**Comment by 2Fwebd** [comment]
Hello,
I updated my 2 dev websites and it worked for 6 extensions. With the website without the update I got the message that we know : Github error Api .... **And on my dev website with your latest release I got this :**
![unyson_ _starry_ _wordpress](https://cloud.githubusercontent.com/assets/5446124/5668865/1b481870-9773-11e4-918f-cef9f282a425.png)
 Any idea ?

**Comment by moldcraft** [comment]
Ops...

I replaced the requests to be done through our custom API cache but didn't notice that the zip url is also a request to the api. 

http://github-api-cache.unyson.io/repos/ThemeFuse/Unyson-Social-Extension/releases

Download link of the latest version is

https://**api.github.com**/repos/ThemeFuse/Unyson-Social-Extension/zipball/v1.0.1

I though it's a direct download link, like on [this page](https://github.com/ThemeFuse/Unyson/releases/tag/v2.1.13) if you click the "**Download Source (zip)**" button, the link is 

https://**github.com**/ThemeFuse/Unyson-Social-Extension/archive/v1.0.1.zip`

----

Sorry about that. I will fix it quick and release a new version.

**Comment by 2Fwebd** [comment]
Hello @moldcraft  !
I think, I understand !! No problem, thank you very much for your AWESOME work !!! Keep me informed and if you need any help to test let me know.

**Comment by moldcraft** [comment]
v2.1.14 is live. Please update and check again if it works.

**Comment by 2Fwebd** [comment]
Thank you for the update. It works sometimes, but most of the time I click one first time and I have the same error than before. **But when I click a second time** (just after) on the button it works (every time). So it's pretty weird no ? But much better than before !!
